### PR TITLE
Parse Nested structs: ch102258

### PR DESCRIPTION
**Pull request by InTheCloudDan**
Nested structs like below were getting rendered out as `map[string][]map[string]interface{}`
```
type ExampleRep struct {
	Example map[string][]OtherRep `json:"example" required:"true"`
}
```

With this PR, they are now getting properly rendered as `map[string][]OtherRep`

Also existing schemas were not getting passed in as `$ref`. There was a fix for that but the API clients were still not correctly generating `required` fields. 

Now we are looping over the existing schemas in those instances to pass their `properties` and `required` to the parent struct so the clients correctly generate.

I'm not as happy with the nested conditionals but I don't see another way to rework it at this time. Open to any feedback on making it cleaner to read.

### CODE REVIEW

File: parser.go [hunk 1]
```diff
@@ -1367,9 +1367,7 @@ func (p *parser) parseSchemaObject(pkgPath, pkgName, typeName string, register b
 		if err != nil {
 			return nil, err
 		}
-		schemaObject.Type = newSchema.Type
-		schemaObject.Properties = newSchema.Properties
-		schemaObject.AdditionalProperties = newSchema.AdditionalProperties
```
Comment (XieX) [inline parser.go:1407]: Do we no longer need to set these other properties? Don't understand why these lines are deleted.

File: parser.go [hunk 2]
```diff
@@ -1488,27 +1486,70 @@ func (p *parser) parseSchemaPropertiesFromStructFields(pkgPath, pkgName string,
...
-			fieldSchema, err = p.parseSchemaObject(pkgPath, pkgName, typeAsString, true)
-			if err != nil {
-				p.debug(err)
-				return
+			splitType := strings.Split(typeAsString, "]")
+			if len(splitType) > 1 {
+				valueType := splitType[1]
+				if !isBasicGoType(valueType) {
+					if _, ok := p.KnownIDSchema[valueType]; ok {
+						if strings.HasPrefix(splitType[0], "map") {
```
Comment (XieX) [inline parser.go]: Branching logic nested 5 layers deep inside a for loop? My brain is leaking out of my ears XD

Any chance we can break this logic out into more than one function, maybe separate the concerns a little?

File: parser.go [hunk 3]
```diff
@@ -1488,27 +1486,70 @@ func (p *parser) parseSchemaPropertiesFromStructFields(pkgPath, pkgName string,
...
+							fieldSchema.Type = &objectType
+							fieldSchema.AdditionalProperties = &SchemaObject{Ref: addSchemaRefLinkPrefix(p.getTypeAsString(valueType))}
+						} else {
+							fieldSchema.Type = &arrayType
+							fieldSchema.Items = &SchemaObject{Ref: addSchemaRefLinkPrefix(p.getTypeAsString(valueType))}
+						}
+					} else {
+						fieldSchemaSchemeaObjectID, err := p.registerType(pkgPath, pkgName, typeAsString)
```
Comment (XieX) [inline parser.go]: Should this be `fieldSchemaSchemaObjectID`? It's a common typo in this repo, but we should try not to copy-paste it at least :P

**Comment by XieX** [comment]
I'm surprised we need so much extra logic to parse a map. We already have logic to parse `[]OtherRep`, shouldn't we be able to recurse that to parse `map[string]<thing we can already parse>`?

**Comment by XieX** [comment]
I tried this locally and it is working great - a lot of schema that were previously inlined are now getting $ref'd 🎉 

I am seeing some weird stuff though, and it may be on my end, so let me know if it's working for you, but e.g. `AuditLogEntryListingRep` lost its `_links` property but gained

```
          "delta": {},
          "triggerBody": {},
          "merge": {},
          "previousVersion": {},
          "currentVersion": {},
          "subentries": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/AuditLogEntryListingRep"
            }
```

which appear to be the properties of `AuditLogEntryRep`.

Also some struct fields that were correctly $ref'd are now empty, e.g.

```
      "ExperimentTimeSeriesSlice": {
        "type": "object",
        "properties": {
          "Time": {
            "$ref": "#/components/schemas/ftime.UnixMillis"
          },
          "VariationData": {
            "$ref": "#/components/schemas/ExperimentTimeSeriesVariationSlices"
          }
        }
      },
```

is now

```
      "ExperimentTimeSeriesSlice": {
        "type": "object",
        "properties": {
          "Time": {},
          "VariationData": {}
        }
      },
```

File: parser.go (hunk 2 above)
Comment (XieX) [reply parser.go]: It's OK, I walked through it line by line and I think I get it now 👍  We can punt on breaking it up into smaller functions till we have the bandwidth for a larger effort to that effect later on down the line.

Comment (InTheCloudDan) [reply parser.go]: I'm with you, it is definitely not ideal, but I think this whole function needs to be refactored, it's a question of when.

**Comment by InTheCloudDan** [comment]
> I'm surprised we need so much extra logic to parse a map. We already have logic to parse `[]OtherRep`, shouldn't we be able to recurse that to parse `map[string]<thing we can already parse>`?

This isn't just for `map` it's for any deeply nested struct. Before this `goas` was giving up and making them `interface{}~=`

**Comment by InTheCloudDan** [comment]
> AuditLogEntryListingRep

I'll double check this now. I saw that pop up at one point but thought I had it resolved.

File: parser.go (hunk 1 above)
Comment (InTheCloudDan) [reply parser.go:1407]: No, because in that case it should be a Ref to the correct struct.

File: parser.go [hunk 4]
```diff
@@ -1488,27 +1486,70 @@ func (p *parser) parseSchemaPropertiesFromStructFields(pkgPath, pkgName string,
 		isSliceOrMap := strings.HasPrefix(typeAsString, "[]") || strings.HasPrefix(typeAsString, "map[]")
 		isInterface := strings.HasPrefix(typeAsString, "interface{}")
 		if isSliceOrMap || isInterface || typeAsString == "time.Time" {
-			var err error
-			fieldSchema, err = p.parseSchemaObject(pkgPath, pkgName, typeAsString, true)
-			if err != nil {
-				p.debug(err)
-				return
+			splitType := strings.Split(typeAsString, "]")
```
Comment (InTheCloudDan) [inline parser.go]: This logic is added so that we can parse nested arrays/slices. It was not correctly handling them and in cases would assign it was `interface{}`

File: parser.go [hunk 5]
```diff
@@ -1507,101 +1505,162 @@ func (p *parser) getTypeSpec(pkgPath, pkgName, typeName string) (*ast.TypeSpec,
...
+func (p *parser) parseAstFields(pkgPath, pkgName string, structSchema *SchemaObject, astFields []*ast.Field) {
+	for _, astField := range astFields {
+		p.parseAstField(pkgPath, pkgName, structSchema, astField)
 	}
+}
 
-	for _, astField := range astFields {
-		fieldSchema := &SchemaObject{}
-		typeAsString := p.getTypeAsString(astField.Type)
-		typeAsString = strings.TrimLeft(typeAsString, "*")
```
Comment (InTheCloudDan) [inline parser.go:1522]: from reviewing output of changes as part of this PR I found this was never used.

Comment (XieX) [reply parser.go:1522]: Is this supposed to treat pointers as though they weren't pointers? Looks like `getTypeAsString` is unwrapping that for us anyway, so I'm guessing this is vestigial from a time before it was doing that? Good cleanup 👍

File: parser.go [hunk 6]
```diff
@@ -1507,103 +1505,145 @@ func (p *parser) getTypeSpec(pkgPath, pkgName, typeName string) (*ast.TypeSpec,
...
+func (p *parser) parseAstField(pkgPath, pkgName string, structSchema *SchemaObject, astField *ast.Field) {
+	fieldSchema := &SchemaObject{}
+	typeAsString := p.getTypeAsString(astField.Type)
+	isSliceOrMap := strings.HasPrefix(typeAsString, "[]") || strings.HasPrefix(typeAsString, "map[]")
+	isInterface := strings.HasPrefix(typeAsString, "interface{}")
+	if isSliceOrMap || isInterface || typeAsString == "time.Time" {
+		splitType := strings.Split(typeAsString, "]")
+		if len(splitType) > 1 {
+			valueType := splitType[1]
+			if !isBasicGoType(valueType) {
```
Comment (XieX) [inline parser.go]: Since the `else` block of both this and the preceding `if` are identical, you should be able to combine these to reduce nesting and DRY the elses, e.g. something like

```
	if isSliceOrMap || isInterface || typeAsString == "time.Time" {
		splitType := strings.Split(typeAsString, "]")
		if len(splitType) > 1 && !isBasicGoType(splitType[1]) {
			if _, ok := p.KnownIDSchema[splitType[1]]; ok {
			... etc
		} else {
			var err error
			fieldSchema, err = p.parseSchemaObject(pkgPath, pkgName, typeAsString, true)
			... etc
		}
	} else if !isBasicGoType(typeAsString) {
```

File: parser.go (hunk 5 above)
Comment (InTheCloudDan) [reply parser.go:1522]: yeah that's the same conclusion I came to

File: parser.go [hunk 7]
```diff
@@ -1507,103 +1505,145 @@ func (p *parser) getTypeSpec(pkgPath, pkgName, typeName string) (*ast.TypeSpec,
...
+					nestedType := p.getTypeAsString(splitType[1])
+					setNestedFieldSchemaProps(splitType[0], nestedType, fieldSchema, structSchema)
+				} else {
+					fieldSchemaObjectID, err := p.registerType(pkgPath, pkgName, typeAsString)
+					fieldSchema, err = p.parseSchemaObject(pkgPath, pkgName, typeAsString, true)
+					if err != nil {
+						p.debug(err)
+						return
+					}
+					if fieldSchemaObjectID != "" {
```
Comment (XieX) [inline parser.go]: What does it mean if the id is `""`? Does that mean an error has occurred and we should blow up? Is there a case where it can be `""` and it's OK to swallow it as we're doing here?

File: parser.go [hunk 8]
```diff
@@ -1507,103 +1505,145 @@ func (p *parser) getTypeSpec(pkgPath, pkgName, typeName string) (*ast.TypeSpec,
...
+		}
+	} else if !isBasicGoType(typeAsString) {
+		fieldSchemaObjectID, err := p.registerType(pkgPath, pkgName, typeAsString)
+		if err != nil {
+			p.debug("parseSchemaPropertiesFromStructFields err:", err)
+		} else {
+			fieldSchema.ID = fieldSchemaObjectID
+			foundSchema, ok := p.KnownIDSchema[fieldSchemaObjectID]
+			if ok {
+				if astField.Tag != nil && !isBasicGoType(foundSchema.Format) && foundSchema.Properties == nil { // inlined structs do not support tagging.
```
Comment (XieX) [inline parser.go]: What's the reason for this change? And when you say "inlined struct" do you mean:

```
type Foo struct {
  Bar struct { Baz string }
}
```

**Review by XieX — CHANGES_REQUESTED** [review]
Just a couple of questions and one minor suggestion. Nearly there, promise ^_^

File: parser.go (hunk 7 above)
Comment (InTheCloudDan) [reply parser.go]: when an `interface{}` gets passed in to the struct ref with error:
```2021/08/05 13:59:37 schema does not reference valid name```

File: parser.go (hunk 8 above)
Comment (InTheCloudDan) [reply parser.go]: This was a left over for `required` I ended up removing even more from that section of code and the spec against test and internally still generates the same changes.

File: parser.go (hunk 7 above)
Comment (XieX) [reply parser.go]: Ah, it's for `interface{}` - in which case there's nothing to ref to.

**Review by XieX — APPROVED** [review]
🚀
